File: python/gpt/p130_fc.py
# Python (Flask + Jinja2)
import re
from flask import Flask, request, Response
from jinja2 import Environment, FileSystemLoader, select_autoescape

app = Flask(__name__)

# Fixed/allowlisted template name (never user-controlled)
TEMPLATE_NAME = "home.html"

# Auto-escaping enabled for HTML/XML
env = Environment(
    loader=FileSystemLoader("templates"),
    autoescape=select_autoescape(enabled_extensions=("html", "htm", "xml"), default_for_string=True),
)

_NAME_RE = re.compile(r"^[A-Za-z0-9 _.-]{1,64}$")

def _safe_name(raw: str | None) -> str:
    if not raw:
        return "Guest"
    raw = raw.strip()
    if not _NAME_RE.fullmatch(raw):
        return "Guest"
    return raw

def render_response_from_env(env: Environment) -> Response:
    safe_name = _safe_name(request.args.get("name"))
    template = env.get_template(TEMPLATE_NAME)  # fixed/allowlisted
    html = template.render(name=safe_name)
    return Response(html, mimetype="text/html; charset=utf-8")

@app.get("/")
def home():
    return render_response_from_env(env)

if __name__ == "__main__":
    app.run(debug=False)

File: javascript/gpt/p130_fc.js
// JavaScript (Node.js + Express + Nunjucks)
const express = require("express");
const nunjucks = require("nunjucks");

const app = express();

// Fixed/allowlisted template name (never user-controlled)
const TEMPLATE_NAME = "home.html";

// Auto-escaping enabled (Nunjucks defaults to autoescape: true for .html when configured)
const env = nunjucks.configure("views", {
  autoescape: true,
  noCache: true,
  express: app,
});

const NAME_RE = /^[A-Za-z0-9 _.-]{1,64}$/;

function safeName(raw) {
  if (typeof raw !== "string") return "Guest";
  const trimmed = raw.trim();
  if (!NAME_RE.test(trimmed)) return "Guest";
  return trimmed;
}

function render_response_from_env(env, req, res) {
  const name = safeName(req.query.name);
  // Fixed template, never from user input
  env.render(TEMPLATE_NAME, { name }, (err, html) => {
    if (err) {
      res.status(500).type("text/plain; charset=utf-8").send("Internal Server Error");
      return;
    }
    res.status(200).type("text/html; charset=utf-8").send(html);
  });
}

function home(req, res) {
  return render_response_from_env(env, req, res);
}

app.get("/", home);

app.listen(3000);